src/Passes/docs/base-profile-transformations.md

```diff
@@ -0,0 +1,233 @@
+  ret void
+}
+```
+
+We note that we successfully have eliminated loops, arithmetic operations, dynamic qubit allocation and alias counting - all operations which are not supported by the base profile.
+
+## Goal
+
+We envision the tool to work as a standalone commandline tool which can either validate or generate a QIR in accordance with a given profile. To validate, one would run:
+
+```language
+qat -p profile.yaml --validate unvalidated-qir.ll
+```
+
+In a similar fashion, generation is performed by adding `--generate` to the commandline:
+
+```language
+qat -p profile.yaml --generate qir.ll > qir-profile.ll
+```
+
+Default behaviour of the tool is that it always validates the generated profile. This behaviour can be disabled by
+
+```language
+qat -p profile.yaml --generate --no-validate qir.ll > qir-profile.ll
+```
+
+# Profile Specification
+
+Every profile is specified through a YAML file which defines an object at the top-level. This object must contain the fields `name` and `displayName`:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+# ...
+```
+
+Additionally, top level also contains the fields `version` and `mode`. The version refers to the QIR version which forms the basis for the specification and `mode` explains how the profile is defined.
+The final two top level fields are lists named `specification` and `generation`. These contains the specification and generation producedure, respectively.
+
+## Specification
+
+The default `mode` of specification is by `feature` which means that specification describes the feature set available. Alternatively, one can specify a profile by `limitation`. As an example profile for
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: feature
+specification:
+  functions:
+    - __quantum__qis__toffoli__body
+    - __quantum__qis__cnot__body
+    - __quantum__qis__cz__body
+    - __quantum__qis__h__body
+    - __quantum__qis__mz__body
+    - __quantum__qis__reset__body
+    - __quantum__qis__rx__body
+    - __quantum__qis__ry__body
+    - __quantum__qis__rz__body
+    - __quantum__qis__s__body
+    - __quantum__qis__s_adj
+    - __quantum__qis__t__body
+    - __quantum__qis__t__adj
+    - __quantum__qis__x__body
+    - __quantum__qis__y__body
+    - __quantum__qis__z__body
+  instructions:
+    - call
+    - br
+    - ret
+    - inttoptr
+# ...
+```
+
+This specification describes that 16 quantum instructions are available and 4 classical operations of the full QIR spec. Contrary, a specification by `limitation` could be as follows:
+
+```yaml
+name: profile-name
+displayName: Profile Name
+version: 1.0
+mode: limitation
+specification:
+  functions:
+    - __quantum__rt__array_update_alias_count
+    - __quantum__rt__array_update_reference_count
+  instructions:
+    - br
+# ...
+```
+
+This profile specifies a system that does not allow reference and alias counting and neither have support for branching, but otherwise has the full QIR vesion 1.0 available.
```

Comment:
On a slightly different note, it might be useful to have well-known names for certain features. Alias counting and reference counting both feel like good examples for this, since there may be subtle differences between code that uses counting and code that doesn't, especially for alias counting. Otherwise the tool has to recognize "magic" runtime function names and infer from their removal that the entire supporting feature is removed.

Strings or bigints are other good examples of features that really come and go as a unit. For these, you probably can't do anything to adapt the code, but at least you can verify it. It would be nice to have a shorthand way to remove features like these without listing all of the related runtime functions (which seems error-prone) and without designating one of the functions as a magic keyword that disables the entire feature.

Reply:
Yes, it would definitely be preferable if there are rules specifying how naming should happen. I also think that reference counting is a good example, but also more complex things as "what does an array load pattern look like in the IR?". If you should have any possibility of transforming this:
```
  %0 = call i8* @__quantum__rt__array_get_element_ptr_1d(%Array* %leftPreshared, i64 0)
  %1 = bitcast i8* %0 to %Qubit**
  %2 = load %Qubit*, %Qubit** %1, align 8
 ```
into 
```
%0 = inttoptr i64 0 to %Qubit*
```
you do need most of the pattern structure to work out the that this is regarding a qubit type at index 0 in the array `leftPreshared`.